contracts/strategy/ConverterStrategyBase.sol

```diff
@@ -0,0 +1,332 @@
+    _withdrawFromPoolUniversal(amount, false);
+  }
+
+  /// @dev Withdraw all from the pool.
+  function _withdrawAllFromPool() override internal virtual {
+    _withdrawFromPoolUniversal(_investedAssets, false);
+  }
+
+  /// @dev If pool support emergency withdraw need to call it for emergencyExit()
+  function _emergencyExitFromPool() override internal virtual {
+    _withdrawFromPoolUniversal(_investedAssets, true);
+  }
+
+
+  function _recycle(address[] memory tokens, uint[] memory amounts) internal {
+    require(tokens.length == amounts.length, "SB: Arrays mismatch");
+    uint len = tokens.length;
+
+    for (uint i = 0; i < len; ++i) {
+      address token = tokens[i];
+      uint amount = amounts[i];
+
+      if (amount > 0) {
+        uint amountToCompound = amount * compoundRatio / COMPOUND_DENOMINATOR;
+        if (amountToCompound > 0) {
+          tetuLiquidator.liquidate(token, asset, amountToCompound, LIQUIDATION_SLIPPAGE);
+        }
+
+        uint amountToForward = amount - amountToCompound;
+        if (amountToForward > 0) {
+          _sendToForwarder(token, amount);
+        }
+      }
+    }
+  }
+
+  /// @dev Claim all possible rewards.
+  function _claim() override internal virtual {
+    address[] memory tokens;
+    uint[] memory amounts;
+
+    (tokens, amounts) = _depositorClaimRewards();
+    _recycle(tokens, amounts);
+
+    (tokens, amounts) = tetuConverter.claimRewards(address(this));
+    _recycle(tokens, amounts);
+  }
+
+  /// @dev Is strategy ready to hard work
+  function isReadyToHardWork()
+  override external view returns (bool) {
+    return _isReadyToHardWork;
+  }
+
+  /// @dev Do hard work
+  function doHardWork()
+  override external returns (uint earned, uint lost) {
+    return _doHardWork(true);
+  }
+
+  /// @dev Do hard work
+  function _doHardWork(bool doDepositToPool)
+  internal returns (uint earned, uint lost) {
+
+    uint assetBalanceBefore = _balance(asset);
+    _claim();
+    earned = _balance(asset) - assetBalanceBefore;
+
+    if (doDepositToPool) {
+      _depositToPool(_balance(asset));
+    }
+
+    lost = 0; // TODO
+
+  }
+
+  // *************************************************************
+  //               OVERRIDES ITetuConverterCallback
+  // *************************************************************
+
+
+  function requireAmountBack (
+    address collateralAsset_,
+    address /*borrowAsset_*/,
+    uint /*requiredAmountBorrowAsset_*/,
+    uint requiredAmountCollateralAsset_
+  ) external override returns (
+    uint amountOut,
+    bool isCollateral
+  ) {
+    _onlyTetuConverter();
+    require(collateralAsset_ == asset, 'CSB: Wrong asset');
+
+    amountOut = 0;
+    uint assetBalance = _balance(collateralAsset_);
+
+    if (assetBalance >=  requiredAmountCollateralAsset_) {
+      amountOut = requiredAmountCollateralAsset_;
+
+    } else {
+      _withdrawFromPool(requiredAmountCollateralAsset_ - assetBalance);
+      amountOut = _balance(collateralAsset_);
```

Comment:
left comment that we assume if withdraw less amount then requiredAmountCollateralAsset_ it will be rebalanced in the next call
or handle this case